src/metabase/data_apps/models.clj

```diff
@@ -0,0 +1,117 @@
+;;------------------------------------------------------------------------------------------------;;
+
+(def ^:private data-app-statuses #{:private :published :archived})
+
+(t2/deftransforms :model/DataApp
+  {:status (mi/transform-validator mi/transform-keyword (partial mi/assert-enum data-app-statuses))})
+
+;;------------------------------------------------------------------------------------------------;;
+;;                                   :model/DataAppDefinition                                     ;;
+;;------------------------------------------------------------------------------------------------;;
+
+(t2/deftransforms :model/DataAppDefinition
+  {:config mi/transform-json})
+
+(mr/def ::AppDefinitionConfig
+  [:maybe [:map {:closed true}
+           [:version pos-int?]]])
+
+(defn- validate-app-definition
+  "Validate an AppDefinition."
+  [app-definition]
+  (mu/validate-throw ::AppDefinitionConfig (:config app-definition)))
+
+(t2/define-before-insert :model/DataAppDefinition
+  [instance]
+  (validate-app-definition instance)
+  instance)
+
+(t2/define-before-update :model/DataAppDefinition
+  [instance]
+  (throw (ex-info "AppDefinition is append-only and cannot be updated"
+                  {:status-code 400
+                   :changes     (t2/changes instance)})))
+
+(defn- next-revision-number
+  "Get the next revision number for a given app."
+  [app-id]
+  (inc (or (:max (t2/select-one [:model/DataAppDefinition [:%max.revision_number :max]] :app_id app-id))
+           0)))
+
+;;------------------------------------------------------------------------------------------------;;
+;;                                    :model/DataAppRelease                                       ;;
+;;------------------------------------------------------------------------------------------------;;
+
+(t2/define-before-insert :model/DataAppRelease
+  [instance]
+  (merge {:published_at :%now} instance))
+
+(t2/define-before-update :model/DataAppRelease
+  [instance]
+  (let [changed-keys (-> instance t2/changes keys set)]
+    (when-not (= #{:retracted} changed-keys)
+      (throw (ex-info (format "AppRelease is append-only. Only 'retracted' field can be updated, but got: %s"
+                              changed-keys)
+                      {:status-code 400}))))
+  instance)
+
+;;------------------------------------------------------------------------------------------------;;
+;;                                       Serialization                                            ;;
+;;------------------------------------------------------------------------------------------------;;
+
+(comment
+  #_todo)
+
+;;------------------------------------------------------------------------------------------------;;
+;;                                        Public APIs                                             ;;
+;;------------------------------------------------------------------------------------------------;;
+
+(defn new-definition!
```

Comment:
nit: I don't like `new` as the "verb" here, how about `update-definition!` or `set-latest-definition!`?